static/src/javascripts/projects/commercial/modules/messenger.ts

```diff
@@ -0,0 +1,409 @@
+	 * to obtain a value of the desired type (or fail gracefully)
+	 */
+	specs: unknown | undefined,
+	/**
+	 * Non-persistent callbacks can be chained together. This value is the return
+	 * value of the previously fired callback in the chain. It is the responsibility
+	 * of the current callback to either ignore it or use it / pass along
+	 */
+	ret: unknown,
+	/**
+	 * Reference to the iframe that is the source of the message
+	 */
+	iframe?: HTMLIFrameElement,
+) => unknown;
+
+/**
+ * The set of listeners currently registered
+ *
+ * Each message sent by an iframe has a `type` field which indicates the kind of
+ * message e.g. `resize`, `measure-ad-load`. One or more listeners is registered
+ * for each type.
+ */
+type Listeners = Partial<
+	Record<MessageType, ListenerCallback | ListenerCallback[] | undefined>
+>;
+
+/**
+ * Types of functions to register a listener for a given type of iframe message
+ */
+export type RegisterListener = (
+	type: MessageType,
+	callback: ListenerCallback,
+	options?: {
+		window?: WindowProxy;
+		persist?: boolean;
+	},
+) => void;
+
+/**
+ * Types of functions to unregister a listener for a given type of iframe message
+ *
+ */
+export type UnregisterListener = (
+	type: MessageType,
+	callback?: ListenerCallback,
+	options?: {
+		window?: WindowProxy;
+	},
+) => void;
+
+const LISTENERS: Listeners = {};
+let REGISTERED_LISTENERS = 0;
+
+const error405 = {
+	code: 405,
+	message: 'Service %% not implemented',
+};
+const error500 = {
+	code: 500,
+	message: 'Internal server error\n\n%%',
+};
+
+/**
+ * Determine if an unknown payload has the shape of a programmatic message
+ *
+ * @param payload The unknown message payload
+ */
+const isProgrammaticMessage = (
+	payload: unknown,
+): payload is ProgrammaticMessage => {
+	const payloadToCheck = payload as ProgrammaticMessage;
+	return (
+		payloadToCheck.type === 'set-ad-height' &&
+		('id' in payloadToCheck.value || 'slotId' in payloadToCheck.value) &&
+		'height' in payloadToCheck.value
+	);
+};
+
+/**
+ * Convert a legacy programmatic message to a standard message
+ *
+ * Note that this only applies to specific resize programmatic messages
+ * (these include specific width and height values)
+ */
+const toStandardMessage = (
+	payload: ProgrammaticMessage,
+): StandardMessage<{ width: number; height: number }> => ({
+	id: 'aaaa0000-bb11-cc22-dd33-eeeeee444444',
+	type: 'resize',
+	iframeId: payload.value.id,
+	slotId: payload.value.slotId,
+	value: {
+		height: +payload.value.height,
+		width: +payload.value.width,
+	},
+});
+
+/**
+ * Retrieve a reference to the calling iframe via it's ID.
+ *
+ * Incoming messages contain the ID of the iframe into which the source window is embedded.
+ */
+const getIframe = (data: StandardMessage): HTMLIFrameElement | undefined => {
+	if (data.slotId) {
+		const container = document.getElementById(`dfp-ad--${data.slotId}`);
+		const iframes = container
+			? container.getElementsByTagName('iframe')
```

Comment:
`getElementsByTagName` returns a live collection - is this required?

Reply:
I don't think so, have gone with your alternative in comment below.